File: frontend/src/app/post.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Post {
    id: number;
    author: string;
    description: string;
    mediaUrl?: string;
    mediaType?: 'IMAGE' | 'VIDEO';
    createdAt: string;
    likeCount: number;
    commentCount: number;
    likedByCurrentUser: boolean;
    hidden : boolean;
    comments?: Comment[];
}

export interface Comment {
    id: number;
    postId: number;
    author: string;
    content: string;
    createdAt: string;
}

@Injectable({
    providedIn: 'root'
})
export class PostService {
    private apiUrl = 'http://localhost:8080/api/v1/posts';
    private interactionUrl = 'http://localhost:8080/api/v1/interactions';

    private http = inject(HttpClient);

    getAllPosts(): Observable<Post[]> {
        return this.http.get<Post[]>(this.apiUrl);
    }

    getFeed(): Observable<Post[]> {
        return this.http.get<Post[]>(`${this.apiUrl}/feed`);
    }

    getUserPosts(username: string): Observable<Post[]> {
        return this.http.get<Post[]>(`${this.apiUrl}/user/${username}`);
    }

    deletePost(id: number): Observable<void> {
        return this.http.delete<void>(`${this.apiUrl}/${id}`);
    }

    createPost(description: string, file?: File): Observable<Post> {
        const formData = new FormData();
        if (description) formData.append('description', description);
        if (file) formData.append('file', file);

        return this.http.post<Post>(this.apiUrl, formData);
    }

    updatePost(id: number, description: string): Observable<Post> {
        return this.http.put<Post>(`${this.apiUrl}/${id}?description=${encodeURIComponent(description)}`, {});
    }

    getComments(postId: number): Observable<Comment[]> {
        return this.http.get<Comment[]>(`${this.interactionUrl}/comments/${postId}`);
    }

    addComment(postId: number, content: string): Observable<Comment> {
        return this.http.post<Comment>(`${this.interactionUrl}/comments/${postId}`, { content });
    }

    likePost(postId: number): Observable<void> {
        return this.http.post<void>(`${this.interactionUrl}/likes/${postId}`, {});
    }

    getLikeCount(postId: number): Observable<number> {
        return this.http.get<number>(`${this.interactionUrl}/likes/${postId}/count`);
    }
}

File: backend/src/main/java/com/example/blog/service/PostService.java
package com.example.blog.service;

import com.example.blog.dto.PostRequest;
import com.example.blog.dto.PostResponse;
import com.example.blog.model.Notification;
import com.example.blog.model.Post;
import com.example.blog.model.User;
import com.example.blog.repository.NotificationRepository;
import com.example.blog.repository.PostRepository;
import com.example.blog.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.io.IOException;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class PostService {

    private final PostRepository postRepository;
    private final FileStorageService fileStorageService;
    private final UserRepository userRepository;
    private final NotificationRepository notificationRepository;

    @Transactional
    public PostResponse createPost(PostRequest request) throws IOException {
        String username = ((UserDetails) SecurityContextHolder.getContext().getAuthentication().getPrincipal())
                .getUsername();
        User user = userRepository.findByUsername(username).orElseThrow();

        String mediaUrl = null;
        String mediaType = null;

        if (request.getFile() != null && !request.getFile().isEmpty()) {
            mediaUrl = fileStorageService.saveFile(request.getFile());
            mediaType = request.getFile().getContentType().startsWith("image") ? "IMAGE" : "VIDEO";
        }

        Post post = Post.builder()
                .author(user)
                .description(request.getDescription())
                .mediaUrl(mediaUrl)
                .mediaType(mediaType)
                .build();
        post = postRepository.save(post);

        // ONLY Notify followers when you post
        try {
            for (User follower : user.getFollowers()) {
                Notification notif = Notification.builder()
                        .user(follower)
                        .message(user.getUsername() + " published a new post!")
                        .read(false)
                        .build();
                notificationRepository.save(notif);
            }
        } catch (Exception e) {
            System.out.println("Could not load followers: " + e.getMessage());
        }

        return mapToResponse(post);
    }

    @Transactional
    public PostResponse updatePost(Long postId, String description) {
        String username = ((UserDetails) SecurityContextHolder.getContext().getAuthentication().getPrincipal())
                .getUsername();
        Post post = postRepository.findById(postId).orElseThrow(() -> new RuntimeException("Post not found"));

        if (!post.getAuthor().getUsername().equals(username)) {
            throw new RuntimeException("Not authorized");
        }

        post.setDescription(description);
        return mapToResponse(postRepository.save(post));
    }

    public List<PostResponse> getAllPosts(boolean isAdmin) {
       List<Post> posts;
    
       if (isAdmin) {
           posts = postRepository.findAllByOrderByCreatedAtDesc();
       } else {
           posts = postRepository.findByHiddenFalseOrderByCreatedAtDesc();
       }
    
       return posts.stream().map(this::mapToResponse).collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<PostResponse> getFeed() {
        String username = SecurityContextHolder.getContext().getAuthentication().getName();
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new RuntimeException("User not found"));
        
        java.util.Set<User> feedAuthors = new java.util.HashSet<>(user.getFollowing());
        feedAuthors.add(user);

        if (feedAuthors.size() == 1) {
            return postRepository.findByHiddenFalseOrderByCreatedAtDesc().stream()
                    .map(this::mapToResponse)
                    .collect(Collectors.toList());
        }

        return postRepository.findByAuthorInAndHiddenFalseOrderByCreatedAtDesc(feedAuthors).stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    public List<PostResponse> getUserPosts(String username) {
        User user = userRepository.findByUsername(username).orElseThrow(() -> new RuntimeException("User not found"));
        return postRepository.findAllByAuthorOrderByCreatedAtDesc(user).stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    public void deletePost(Long postId) {
        String username = ((UserDetails) SecurityContextHolder.getContext().getAuthentication().getPrincipal())
                .getUsername();
        Post post = postRepository.findById(postId).orElseThrow(() -> new RuntimeException("Post not found"));

        if (!post.getAuthor().getUsername().equals(username)) {
            throw new RuntimeException("Not authorized");
        }

        postRepository.delete(post);
    }

    private PostResponse mapToResponse(Post post) {
        return PostResponse.builder()
                .id(post.getId())
                .author(post.getAuthor().getUsername())
                .description(post.getDescription())
                .mediaUrl(post.getMediaUrl())
                .mediaType(post.getMediaType())
                .createdAt(post.getCreatedAt())
                .likeCount(post.getLikeCount())
                .commentCount(post.getCommentCount())
                .build();
    }
}